xarray/namedarray/utils.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
+
+import importlib
+import sys
+import typing
+from collections.abc import Iterator, Mapping
+
+import numpy as np
+
+if typing.TYPE_CHECKING:
+    if sys.version_info >= (3, 10):
+        from typing import TypeGuard
+    else:
+        from typing_extensions import TypeGuard
+
+# temporary placeholder for indicating an array api compliant type.
+# hopefully in the future we can narrow this down more
+T_DuckArray = typing.TypeVar("T_DuckArray", bound=typing.Any)
+
+K = typing.TypeVar("K")
+V = typing.TypeVar("V")
+T = typing.TypeVar("T")
+
+
+class Frozen(Mapping[K, V]):
+    """Wrapper around an object implementing the mapping interface to make it
+    immutable. If you really want to modify the mapping, the mutable version is
+    saved under the `mapping` attribute.
+    """
+
+    __slots__ = ("mapping",)
+
+    def __init__(self, mapping: Mapping[K, V]):
+        self.mapping = mapping
+
+    def __getitem__(self, key: K) -> V:
+        return self.mapping[key]
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self.mapping)
+
+    def __len__(self) -> int:
+        return len(self.mapping)
+
+    def __contains__(self, key: object) -> bool:
+        return key in self.mapping
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}({self.mapping!r})"
+
+
+def module_available(module: str) -> bool:
+    """Checks whether a module is installed without importing it.
+
+    Use this for a lightweight check and lazy imports.
+
+    Parameters
+    ----------
+    module : str
+        Name of the module.
+
+    Returns
+    -------
+    available : bool
+        Whether the module is installed.
+    """
+    return importlib.util.find_spec(module) is not None
+
+
+def is_dask_collection(x: typing.Any) -> bool:
+    if module_available("dask"):
+        from dask.base import is_dask_collection
+
+        return is_dask_collection(x)
+    return False
+
+
+def is_duck_array(value: typing.Any) -> TypeGuard[T_DuckArray]:
+    if isinstance(value, np.ndarray):
+        return True
+    return (
+        hasattr(value, "ndim")
+        and hasattr(value, "shape")
+        and hasattr(value, "dtype")
+        and (
+            (hasattr(value, "__array_function__") and hasattr(value, "__array_ufunc__"))
+            or hasattr(value, "__array_namespace__")
+        )
+    )
```

Comment:
Why should both `__array_function__` and `__array_ufunc__` exist?

This does not align with `as_compatible_data` where only function is needed.

In the tests we seem to assume only `__array_function__` is necessary.

Reply:
> This does not align with as_compatible_data where only function is needed.
> In the tests we seem to assume only __array_function__ is necessary.

after reading this comment from stephan, i'm unsure about the reason behind this. Is it possible that this was unintentional? 

> Right now people often check for the existence of `__array_function__` to indicate NumPy duck arrays, which clearly isn't quite right. -- https://github.com/pydata/xarray/issues/3917#issuecomment-611323478



Cc @shoyer & @dcherian

Reply:
I think the check `as_compatible_data` has stayed pretty much unchanged ever since its original addition, so I would think that that's the case. I would replace the custom check with `is_duck_array`, but that would pass through `ndarray` subclasses (might be worth considering since it would allow wrapping things like `astropy.units`, but it might also need a bit more discussion).